File: Winter2022/examples/java/Lecture 6/ThreadingHandlerRunnable/app/src/main/java/mobiledev/unb/ca/threadinghandlerrunnable/LoadIconTask.java
package mobiledev.unb.ca.threadinghandlerrunnable;

import android.content.Context;
import android.graphics.BitmapFactory;
import android.os.Handler;
import android.os.Looper;
import android.widget.ImageView;
import android.widget.ProgressBar;

public class LoadIconTask extends Thread {
    private final int mBitmapResID = R.drawable.painter;
    private final Handler mHandler;
    private ImageView mImageView;
    private ProgressBar mProgressBar;
    private final Context mAppContext;

    LoadIconTask(Context context) {
        mAppContext = context;
        mHandler = new Handler(Looper.getMainLooper());
    }

    public LoadIconTask setImageView(ImageView mImageView) {
        this.mImageView = mImageView;
        return this;
    }

    public LoadIconTask setProgressBar(ProgressBar mProgressBar) {
        this.mProgressBar = mProgressBar;
        return this;
    }

    public void run() {
        mHandler.post(() -> mProgressBar.setVisibility(ProgressBar.VISIBLE));

        // Simulating long-running operation
        for (int i = 1; i < 11; i++) {
            sleep();
            final int step = i;
            mHandler.post(() -> mProgressBar.setProgress(step * 10));
        }

        mHandler.post(() -> mImageView.setImageBitmap(
                BitmapFactory.decodeResource(mAppContext.getResources(), mBitmapResID)));

        mHandler.post(() -> mProgressBar.setVisibility(ProgressBar.INVISIBLE));
    }

    private void sleep() {
        try {
            int mDelay = 500;
            Thread.sleep(mDelay);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}

File: Fall2023/examples/Lecture 6/ThreadingHandlerRunnable/app/src/main/java/mobiledev/unb/ca/threading/handlerrunnable/LoadIconTask.kt
package mobiledev.unb.ca.threading.handlerrunnable

import android.content.Context
import android.widget.ProgressBar
import android.graphics.BitmapFactory
import android.os.Handler
import android.os.Looper
import android.widget.ImageView

class LoadIconTask internal constructor(private val mAppContext: Context) : Thread() {
    private val mHandler: Handler = Handler(Looper.getMainLooper())
    private var imageView: ImageView? = null
    private var progressBar: ProgressBar? = null
    private val mBitmapResID = R.drawable.painter

    fun setImageView(imageView: ImageView?): LoadIconTask {
        this.imageView = imageView
        return this
    }

    fun setProgressBar(progressBar: ProgressBar?): LoadIconTask {
        this.progressBar = progressBar
        return this
    }

    override fun run() {
        mHandler.post { progressBar!!.visibility = ProgressBar.VISIBLE }

        // Simulating long-running operation
        for (i in 1..10) {
            sleep()
            mHandler.post { progressBar!!.progress = i * 10 }
        }
        mHandler.post {
            imageView!!.setImageBitmap(
                BitmapFactory.decodeResource(mAppContext.resources, mBitmapResID))
        }
        mHandler.post { progressBar!!.visibility = ProgressBar.INVISIBLE }
    }

    private fun sleep() {
        try {
            val mDelay = 500
            sleep(mDelay.toLong())
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

}